**Project 7 - Detecting Fake News with Python and Machine Learning**

**What is Fake News?**

एक प्रकार की पीत पत्रकारिता, नकली समाचार उन समाचारों के अंशों को समाहित करती है जो धोखा हो सकते हैं और आम तौर पर सोशल मीडिया और अन्य ऑनलाइन मीडिया के माध्यम से फैलाए जाते हैं। यह अक्सर कुछ विचारों को आगे बढ़ाने या थोपने के लिए किया जाता है और अक्सर राजनीतिक एजेंडे के साथ हासिल किया जाता है। ऐसे समाचार आइटम में झूठे और/या अतिरंजित दावे हो सकते हैं, और एल्गोरिदम द्वारा वायरल किए जा सकते हैं, और उपयोगकर्ता फ़िल्टर बबल में समाप्त हो सकते हैं।

**What is a TfidfVectorizer?**

TF (Term Frequency): The number of times a word appears in a document is its Term Frequency. A higher value means a term appears more often than others, and so, the document is a good match when the term is part of the search terms.

**IDF (Inverse Document Frequency):** 

Words that occur many times a document, but also occur many times in many others, may be irrelevant. IDF is a measure of how significant a term is in the entire corpus.

The TfidfVectorizer converts a collection of raw documents into a matrix of TF-IDF features.

**What is a PassiveAggressiveClassifier?**

Passive Aggressive algorithms are online learning algorithms. Such an algorithm remains passive for a correct classification outcome, and turns aggressive in the event of a miscalculation, updating and adjusting. Unlike most other algorithms, it does not converge. Its purpose is to make updates that correct the loss, causing very little change in the norm of the weight vector.

In [1]:
# importing required libraries

import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

Reading the data into a DataFrame




In [15]:
#Reading the data
df = pd.read_csv('news.csv', error_bad_lines=False, engine="python")

# shape and head
df.shape
df.head()

/usr/local/lib/python3.8/dist-packages/IPython/core/interactiveshell.py:3326: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version.


  exec(code_obj, self.user_global_ns, self.user_ns)
Skipping line 2396: unexpected end of data


Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

Getting the labels from the DataFrame

In [16]:
# Getting the labels
labels=df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

 Spliting the dataset into training and testing sets

In [17]:
 #Spliting the dataset
x_train,x_test,y_train,y_test=train_test_split(df['text'], labels, test_size=0.2, random_state=7)

Let’s initialize a TfidfVectorizer with stop words from the English language and a maximum document frequency of 0.7 (terms with a higher document frequency will be discarded). Stop words are the most common words in a language that are to be filtered out before processing the natural language data. And a TfidfVectorizer turns a collection of raw documents into a matrix of TF-IDF features.

Now, fit and transform the vectorizer on the train set, and transform the vectorizer on the test set.

In [18]:
# Initialize a TfidfVectorizer
tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)

#Fit and transform train set, transform test set
tfidf_train=tfidf_vectorizer.fit_transform(x_train) 
tfidf_test=tfidf_vectorizer.transform(x_test)

Now, initialize a PassiveAggressiveClassifier. We’ll fit this on tfidf_train and y_train.

And after that, we’ll predict on the test set from the TfidfVectorizer and calculate the accuracy with accuracy_score() from sklearn.metrics.

In [19]:
#Initialize a PassiveAggressiveClassifier
pac=PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train,y_train)

#Predict on the test set and calculate accuracy
y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 90.4%


Here, we are getting an accuracy of 90.4% with this model. Finally, let’s print out a confusion matrix to gain insight into the number of false and true negatives and positives.

In [20]:
#print confusion matrix
confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])

array([[246,  23],
       [ 23, 187]])

So with this model, we have 246 true positives, 187 true negatives, 23 false positives, and 23 false negatives.



**Summary**

Today, we learned to detect fake news with Python. We took a political dataset, implemented a TfidfVectorizer, initialized a PassiveAggressiveClassifier, and fit our model. We ended up obtaining an accuracy of 90.4% in magnitude.

Hope you enjoyed the fake news detection python project. 